sql/src/main/java/io/crate/operation/join/nestedloop/UnsortedNestedLoopStrategy.java

```diff
@@ -0,0 +1,137 @@
+
+import com.google.common.base.Optional;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import io.crate.executor.PageInfo;
+import io.crate.executor.TaskResult;
+import io.crate.operation.projectors.Projector;
+
+import javax.annotation.Nonnull;
+import java.util.concurrent.Executor;
+
+public class UnsortedNestedLoopStrategy extends OneShotNestedLoopStrategy {
+
+    public UnsortedNestedLoopStrategy(NestedLoopOperation nestedLoopOperation,
+                                      NestedLoopExecutorService nestedLoopExecutorService) {
+        super(nestedLoopOperation, nestedLoopExecutorService);
+    }
+
+    @Override
+    public String name() {
+        return "unsorted nl";
+    }
+
+    @Override
+    public NestedLoopExecutor executor(JoinContext ctx, Optional<PageInfo> pageInfo, Projector downstream, FutureCallback<Void> callback) {
+        return new UnsortedNestedLoopExecutor(ctx, pageInfo, downstream, nestedLoopExecutorService.executor(), callback);
+    }
+
+    private static class UnsortedNestedLoopExecutor implements NestedLoopExecutor {
+
+        private final JoinContext ctx;
+        private final Optional<PageInfo> globalPageInfo;
+        private final Projector downstream;
+        private final Executor nestedLoopExecutor;
+        private final FutureCallback<Void> finalCallback;
+        private final FutureCallback<TaskResult> onInnerPage;
+        private final FutureCallback<TaskResult> onOuterPage;
+
+        public UnsortedNestedLoopExecutor(JoinContext ctx,
+                                          Optional<PageInfo> globalPageInfo,
+                                          Projector downstream,
+                                          Executor nestedLoopExecutor,
+                                          final FutureCallback<Void> finalCallback) {
+            this.ctx = ctx;
+            this.globalPageInfo = globalPageInfo;
+            this.downstream = downstream;
+            this.nestedLoopExecutor = nestedLoopExecutor;
+            this.finalCallback = finalCallback;
+
+            onInnerPage = new FutureCallback<TaskResult>() {
+                @Override
+                public void onSuccess(TaskResult result) {
+                    onNewInnerPage(result);
+                }
+
+                @Override
+                public void onFailure(@Nonnull Throwable t) {
+                    finalCallback.onFailure(t);
+                }
+            };
+
+            onOuterPage = new FutureCallback<TaskResult>() {
+                @Override
+                public void onSuccess(TaskResult result) {
+                    onNewOuterPage(result);
+                }
+
+                @Override
+                public void onFailure(@Nonnull Throwable t) {
+                    finalCallback.onFailure(t);
+                }
+            };
+        }
+
+        private void joinOuterPage() {
+            if (ctx.innerIsFinished()) {
```

Comment:
this is only true if innerRelation has been iterated, in that case you can do ctx.resetInnerPageInfo() instead of advancing it